Allow accurate even sharing of PV costs for Pods that share PVCs

kaelanspatel opened the pull request:
## What does this PR change?
Implements a modified sweep-line algorithm and associated structs to share PV costs in the instance where multiple pods have the same PVC. Also adds associated tests.

Information about why this was necessary, as well as the workings of the implementation in this PR can be found in [this doc](https://docs.google.com/document/d/1F2P-CMJNnS1uiMXuE0XmwHDeZG3reX8BKKD7UNmsX6Q/edit#).

## Does this PR rely on any other PRs?
No.


## How does this PR impact users? (This is the kind of thing that goes in release notes!)
Fix an issue where PV costs were being duplicated for shared PVCs.

## Links to Issues or ZD tickets this PR addresses or fixes

- https://github.com/kubecost/cost-analyzer-helm-chart/issues/1040
- 


## How was this PR tested?
Wrote tests for each individual function, as well as manually tested on multiple clusters both with pods sharing PVCs and not.

## Have you made an update to documentation?
Not yet, but they should be forthcoming assuming this is approved.

nikovacevic commented on pkg/costmodel/allocation_helpers.go:
```diff
@@ -0,0 +1,133 @@
+		// If the current point happens at a later time than the previous point
+		if !point.Time.Equal(prevPoint.Time) {
+			for key := range activeKeys {
+				// Add a two-long slice of float64 representing:
+				// 1. The division of cost based on how many pods were running between those points
+				// 2. The ratio of the time between those points to the total time that pod was running
```
See comment below about whether this could/should be a struct or not.

nikovacevic commented on pkg/costmodel/allocation_helpers.go:
```diff
@@ -0,0 +1,133 @@
+		point := intervals[i]
+		prevPoint := intervals[i-1]
+
+		// intervals will always have at least two IntervalPoints (one start/end)
+		if i == 1 {
+			activePods += 1
+			activeKeys[prevPoint.Key] = struct{}{}
+		}
```
Is this meant to be the initialization? If so, what do you think about doing this outside the `for` loop so that we don't check this condition each time?

Something like this, above the loop:
```go
// Track active state, starting after the first point (i.e. first pod) has been recorded.

// Number of currently active pods, tracked through the sweep
activePods := 1.0

// Set of keys from all points that have reached "start" but not yet "end"
activeKeys := map[podKey]struct{}{
  intervals[0].prevPoint.Key: struct{}{},
}
```

kaelanspatel replied:
This is a little hacky, but essentially it's adding a map entry with a zero memory size value just to act as a sort of "record" for what specific pods are active at that point in the loop, because each of those pods must then [get a piece of their coefficient appended](https://github.com/kubecost/cost-model/pull/1036/files/5328ab40e1868727b00c7237a913e54f411d3d2b#diff-fb9d5089e2629c9bd167b04b7a80f1d70db00e805dabeed463be0913fd101e8aR90) to their key in the result map.

This could be done with a slice, but since elements need to be deleted out of order, I used a map.

nikovacevic replied:
Right, totally with you on the purpose. My question was meant to be about how you're initializing it. As written, this checks for initialization every single time through the loop. Proposal was only to make the initial condition explicit above the loop so we don't have to check `if i == 1` repeatedly.

kaelanspatel replied:
Ack, you're absolutely correct, I wasn't looking at the right part of the code.

In that case, totally agree.

kaelanspatel replied:
Implemented this.

nikovacevic commented on pkg/costmodel/allocation_helpers.go:
```diff
@@ -0,0 +1,133 @@
+				// 2. The ratio of the time between those points to the total time that pod was running
+				pvcCostCoefficientMap[key] = append(
+					pvcCostCoefficientMap[key],
+					[]float64{1.0 / activePods, point.Time.Sub(prevPoint.Time).Minutes() / pvcIntervalMap[key].Duration().Minutes()},
```
Also, do you think it's worth `NaN` checking here? (Both values could be `NaN`, technically.)

It's obviously unlikely that `pvcIntervalMap[key].Duration().Minutes()` will be zero, but you never really know what's possible with Prometheus data, and `NaN`s tend to spread unhappiness.

kaelanspatel replied:
I think `NaN` checking is probably a good idea. 

I considered creating a new struct here instead of using slices, but thought it would be a little too cluttered for what is essentially a broken-down addition problem that is turned into a single number in short order. It certainly would make a lot more sense visually to be able to see something like:

```
CoefficientComponent{
	Time: point.Time.Sub(prevPoint.Time).Minutes()
	Proportion: 1.0 / activePods
}
```

so I'll try implementing it. I'd have used a tuple if they existed.

nikovacevic replied:
Makes complete sense. Naming a new struct does make for an extra task, which I appreciate. But the two downsides of using a slice are:

(1) making use of that data structure in the future requires someone to go back and figure out what `[][]float64` means -- are there always a set number of entries? what do they mean? and

(2) Go will be more economical allocating memory to a well-defined struct than to a variable-length slice

kaelanspatel replied:
Implemented the `CoefficientComponent` type. Hopefully it's clearer.

nikovacevic commented on pkg/costmodel/allocation_helpers.go:
```diff
@@ -0,0 +1,133 @@
+}
+
+// sortIntervalPoints sorts a list of IntervalPoints from earliest
+// to latest IntervalPoint.Time. In the case that two IntervalPoints
+// have the same time, the point with Type "start" is treated as coming
+// before the "end".
```
Is this condition a requirement for the algorithm to work? Just curious.

kaelanspatel replied:
Hm, not in this case. I guess when I was prototyping this I was thinking that this would count as an overlap just by nature of the segments touching, but that's obviously not true when talking about proportion of cost.

nikovacevic commented on pkg/costmodel/allocation_helpers.go:
```diff
@@ -0,0 +1,133 @@
+
+	for i := range coefficientComponents {
+
+		cost := coefficientComponents[i][0]
```
At this point, do we expect this value (`coefficientComponents[i][0]`) to be the cost? Or is it the percentage of all pods connected to the PV? Trying to figure out if I missed where cost might have been factored into these `coefficientComponents`.

nikovacevic replied:
Also -- not a huge deal, but `getCoefficient` is a relatively vague name for `package costmodel`. We might want to come up with a name more specific to this precise use-case.

kaelanspatel replied:
Yeah, I'll see if I can clarify both the name of this function and (by adding a struct) the coefficient component items.

I really should not have named that cost. You are correct that it is the percentage of all pods connected to the PV. Totally my bad, I think I was thinking cost as in "proportion of cost".

nikovacevic commented on pkg/costmodel/allocation_helpers_test.go:
```diff
@@ -0,0 +1,336 @@
+	"github.com/kubecost/cost-model/pkg/kubecost"
+)
+
+func TestGetIntervalPointsFromWindows(t *testing.T) {
```
Love these tests!

nikovacevic commented on pkg/costmodel/allocation_helpers_test.go:
```diff
@@ -0,0 +1,336 @@
+				NewIntervalPoint(time.Date(2021, 2, 19, 9, 0, 0, 0, time.UTC), "end", podKey{Pod: "Pod1"}),
+			},
+		},
+	}
```
Do you think it's worth testing cases where there are no pods? (Like, maybe the PVC exists, but the pod can't be scheduled.) Or a gap between pods?

I wonder if that might generate `NaN` values because `activePods` could go to `0.0`.

kaelanspatel replied:
I'll take a look, but off the top of my head these are probably covered just because of the existing loop conditions. 

For no pods, it would mean no `Interval` objects would be generated, because we're iterating pods to get those intervals (and, ofc, to even set PV costs regardless). 

For `activePods` going to `0.0`, that's actually fine, because `activePods` is only referenced in a loop that iterates `activeKeys`, which should have a parity `len(activeKeys) == activePods`. 

Now that I've said that, activePods isn't strictly necessary as a separate variable....

nikovacevic reviewed (commented):
Lots of feedback here, so please let me know if you want to discuss anything! Otherwise, just ping me when you're ready for a second pass.

kaelanspatel commented on pkg/costmodel/allocation.go:
```diff
@@ -375,6 +375,68 @@ func (cm *CostModel) ComputeAllocation(start, end time.Time, resolution time.Dur
+
+	for _, pod := range podMap {
+
+		for _, alloc := range pod.Allocations {
```
That's correct. This is essentially just the code which we already were using to get PV hours, just moved into its own loop because I needed that info before setting the cost.

Looking back on it, I see that it's definitely unnecessary to go through every allocation, though to get to `alloc.Properties` I would still need to at least look at one.

kaelanspatel replied:
This should now only look at a single allocation within `pod.Allocations`. Not an enormous change, but accessing the keys of `pod.Allocation` would take an additional loop over containers to begin with, so this just cuts down the number of iterations of this loop to one.

kaelanspatel commented on pkg/costmodel/allocation.go:
```diff
@@ -423,6 +485,13 @@ func (cm *CostModel) ComputeAllocation(start, end time.Time, resolution time.Dur
+					if coeffComponents, ok := sharedPVCCostCoefficientMap[pvcKey][podKey]; ok {
+						cost *= getCoefficient(coeffComponents)
+					} else {
+						klog.Warningf("CostModel.ComputeAllocation: allocation %s and PVC %s have relation but no coeff", alloc.Name, pvc.Name)
```
No, that's just what was being used in the rest of the file. Do we want to use our own package over `klog` in general? I'm not really familiar with the pros and cons.

nikovacevic replied:
Yeah, whenever possible I think we should use our package for consistency. (That file has 3 remaining legacy instances of `klog`, but about 70 of `log`.) If there are drawbacks, that's another story obviously.

kaelanspatel replied:
Replaced klog with log, including the occurrences in other functions.

kaelanspatel commented on pkg/costmodel/allocation.go:
```diff
@@ -375,6 +375,68 @@ func (cm *CostModel) ComputeAllocation(start, end time.Time, resolution time.Dur
+
+		// Get single-point intervals from alloc-PVC relation windows.
+		intervals := getIntervalPointsFromWindows(podIntervalMap)
+		pvcCostCoefficientMap := make(map[podKey][][]float64)
+
+		// Determine coefficients for each PVC-pod relation.
+		getPVCCostCoefficients(intervals, podIntervalMap, pvcCostCoefficientMap)
+
+		sharedPVCCostCoefficientMap[pvcKey] = pvcCostCoefficientMap
```
That sounds reasonable, as this would also address the issue where `pvcCostCoefficientMap` is undefined as you noted in `allocation_helpers.go`.

kaelanspatel replied:
Made `getPVCCostCoefficients` return a value of `map[podKey][]CoefficientComponent` instead of mutating an externally defined value.

kaelanspatel commented on pkg/costmodel/allocation_helpers.go:
```diff
@@ -0,0 +1,133 @@
+package costmodel
```
`intervals.go` sounds more descriptive, I'll use that.

kaelanspatel commented on pkg/costmodel/allocation_helpers.go:
```diff
@@ -0,0 +1,133 @@
+
+}
+
+// sortIntervalPoints sorts a list of IntervalPoints from earliest
+// to latest IntervalPoint.Time. In the case that two IntervalPoints
+// have the same time, the point with Type "start" is treated as coming
+// before the "end".
+func sortIntervalPoints(intervals []IntervalPoint) {
+	sort.Slice(intervals, func(i, j int) bool {
+		if intervals[i].Time.Equal(intervals[j].Time) {
+			return intervals[i].PointType == "start" && intervals[j].PointType == "end"
+		}
+		return intervals[i].Time.Before(intervals[j].Time)
+	})
+}
```
Sounds good, I'll look into it.

kaelanspatel replied:
I believe this has been implemented, though please double check I've understood correctly.

kaelanspatel commented on pkg/costmodel/allocation_helpers.go:
```diff
@@ -0,0 +1,133 @@
+				// Add a two-long slice of float64 representing:
+				// 1. The division of cost based on how many pods were running between those points
+				// 2. The ratio of the time between those points to the total time that pod was running
+				pvcCostCoefficientMap[key] = append(
```
Absolutely, thanks for the insight!

kaelanspatel commented on pkg/costmodel/allocation_helpers.go:
```diff
@@ -0,0 +1,133 @@
+
+	activeKeys := make(map[podKey]struct{})
+
+	// For each interval i.e. for any time a pod-PVC relation ends or starts...
+	for i := 1; i < len(intervals); i++ {
+
+		point := intervals[i]
+		prevPoint := intervals[i-1]
```
Yes, the intervals are generated through a `window` struct's start/end fields, and there can only be one start and one end. The only other alternative is no window, which means the loop doesn't run.

kaelanspatel commented on pkg/costmodel/allocation_helpers.go:
```diff
@@ -0,0 +1,133 @@
+			activeKeys[prevPoint.Key] = struct{}{}
+		}
+
+		// If the current point happens at a later time than the previous point
+		if !point.Time.Equal(prevPoint.Time) {
```
Yes. Originally I had the sorting happen inside this function, but for testing purposes I moved it into the function that generates `[]IntervalPoint` from `map[podKey]kubecost.Window`.

I suppose ideally it should be in neither and have its own test.

kaelanspatel commented:
Updated implementation with:
- Cleanup of certain messy aspects (unnecessary conditions inside loops, extraneous loop iterations, etc).
- Renamed files containing interval logic/tests to `intervals.go` and `intervals_test.go`.
- Replaced klog with kubecost log package.
- Coefficient component values are now stored in the `CoefficientComponent` type, rather than in a length 2 `[]float64`.
- Implemented `sort.Sort` for slices of `IntervalPoints`.

@nikovacevic thanks for all your feedback! I'd love another lookover.

kbrwn commented:
@nikovacevic @kaelanspatel is there any update on this PR? There are a few customers waiting for this.

AjayTripathy commented:
I've tagged this with next-release. @kaelanspatel let's make sure we can get this in soon...

kaelanspatel commented:
Sure. I know @nikovacevic wanted to take another look at this but I'm pretty confident I've addressed his original feedback. The logic remains the same, so if he is ok with merging this I think it's ready to go.

nikovacevic reviewed (approved):
Thank you @kaelanspatel!
